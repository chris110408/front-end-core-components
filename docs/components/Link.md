# Link

The Link component is intended to be used alongside the [Language](Language.md) state. It can be used without any language state although at that point it's redundant. It is not directly dependent on the [Language](Language.md) component and is compatible with other solutions via it's own Provider (See [usage](#usage) below for details).

## Dependencies

* react `^16.3`
* prop-types `^15.6.2`

## Reference

### Provider

| prop name | description | type | required? | default value |
| --------- | ----------- | ---- | --------- | ------------- |
| lang | Current language in the URL | `String` | ✗ | - |
| isStatic | Callback to check if a URL is static, accepts single argument of href (String) and should return a Boolean | `Function` | ✗ | (href) => false |
| children | Child nodes | `Node` | ✓ | - |
| component | The component to use to render the Link | `String\|Function` | ✗ | a |

### Link

| prop name | description | type | required? | default value |
| --------- | ----------- | ---- | --------- | ------------- |
| href | The URL for the link with or without a language prefix | `String` | ✓ | - |
| children | Child nodes | `Node` | ✓ | - |

Any other props are forwarded to the component defined on the Provider.

## Usage

### Provider (optional)

Put the provider high up in your application component tree, usually within `App.js` or the main entry point to your application/website.

```javascript
import React from 'react'
import { Provider } from '@input-output-hk/front-end-core-components/components/Link'

export default () => (
  // The lang prop and the isStatic prop are completely optional.
  // By default the isStatic prop will be a function returning false
  <Provider lang='' isStatic={(href) => {
    // check if href is static for your use case
    return false
  }}>
    ...
    <p>The rest of your app</p>
    ...
  </Provider>
)

```

### Link (consumes data from Provider where provided)

#### Basic example

The example below is the most basic and simply utilises the Link feature of transforming the href.

```javascript
import React from 'react'
import Link from '@input-output-hk/front-end-core-components/components/Link'

export default () => (
  // This will render the component (by default <a />) with the
  // normalized href
  <Link href='/'>
    Language prefixed href when setup for language support
  </Link>
)

```

## Example using Language state

An example making use of the [Language](Language.md) state, as well as implementing a solution for `isStatic`.

### Provider

```javascript
import React from 'react'
import Language from '@input-output-hk/front-end-core-components/components/Language'
import { Provider } from '@input-output-hk/front-end-core-components/components/Link'

export default () => (
  <Language.Consumer>
    {({ lang }) => (
      <Provider lang={lang} isStatic={(href) => {
        // Any href starting with /static/ is assumed a static href
        if (href.match(/^\/static\//)) return true
        return false
      }}>
        ...
        <p>The rest of your app</p>
        ...
      </Provider>
    )}
  </Language.Consumer>
)

```
