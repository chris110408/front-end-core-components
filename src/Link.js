import React, { forwardRef, createContext } from 'react'
import PropTypes from 'prop-types'
import { isRelative } from './lib/url'

const LinkContext = createContext()

export const Provider = ({ children, component = 'a', isStatic = (href) => false, lang = '' }) => (
  <LinkContext.Provider
    value={{ lang, isStatic, component }}
  >
    {children}
  </LinkContext.Provider>
)

Provider.propTypes = {
  children: PropTypes.node.isRequired,
  lang: PropTypes.string,
  isStatic: PropTypes.func,
  component: PropTypes.oneOfType([ PropTypes.string, PropTypes.func ])
}

const LinkInner = (props) => {
  // Transform original href to prefix language where applicable
  function getHref () {
    if (props.lang && isRelative(props.href) && !props.isStatic(props.href) && props.href.substring(1, props.lang.length + 1) !== props.lang) return `/${props.lang}${props.href}`
    return props.href
  }

  const Component = props.component
  const componentProps = {
    ...props,
    href: getHref(),
    isStatic: props.isStatic(props.href),
    isRelative: isRelative(props.href)
  }

  delete componentProps.component
  if (typeof Component === 'string') {
    delete componentProps.isStatic
    delete componentProps.isRelative
  }

  return (
    <Component {...componentProps} />
  )
}

const Link = forwardRef((props, ref) => (
  <LinkContext.Consumer>
    {({ lang, isStatic = () => false, component = 'a' } = {}) => (
      <LinkInner {...props} ref={ref} lang={lang} isStatic={isStatic} component={component} />
    )}
  </LinkContext.Consumer>
))

LinkInner.propTypes = {
  href: PropTypes.string.isRequired,
  lang: PropTypes.string,
  isStatic: PropTypes.func,
  ref: PropTypes.any,
  component: PropTypes.oneOfType([ PropTypes.string, PropTypes.func ])
}

Link.propTypes = {
  href: PropTypes.string.isRequired
}

export default Link
