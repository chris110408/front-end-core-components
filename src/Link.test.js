import React from 'react'
import Link, { Provider } from './Link'
import { mount } from 'enzyme'

const DEFAULT_PROPS = {
  href: '/'
}

function getWrapper ({ props = {}, providerProps = {} } = {}) {
  return mount(
    <Provider {...providerProps}>
      <Link
        {...DEFAULT_PROPS}
        {...props}
      >
        {({ href, isStatic, isRelative }) => (
          <a data={{ isStatic, isRelative }} href={href}>
            A link
          </a>
        )}
      </Link>
    </Provider>
  )
}

describe('<Link />', () => {
  test('it renders correctly with minimal props', () => {
    expect(getWrapper()).toMatchSnapshot()
  })

  describe('when the context provider has a lang set', () => {
    let lang
    beforeEach(() => {
      lang = 'en'
    })

    test('it prefixes the URL with the lang', () => {
      expect(getWrapper({ providerProps: { lang }, props: { href: '/page/' } })).toMatchSnapshot()
    })

    test('already language prefixed URL\'s are unaffected', () => {
      expect(getWrapper({ providerProps: { lang }, props: { href: '/en/page/' } })).toMatchSnapshot()
    })

    test('absolute URL\'s are unaffected', () => {
      expect(getWrapper({ providerProps: { lang }, props: { href: 'https://site.com/page/' } })).toMatchSnapshot()
    })

    describe('when validation for static URL\'s are in place', () => {
      let isStatic
      beforeEach(() => {
        isStatic = (href) => {
          if (href.match(/\.jpeg$/i)) return true
          return false
        }
      })

      test('static URL\'s are unaffected', () => {
        expect(getWrapper({ providerProps: { lang, isStatic }, props: { href: '/images/myimage.jpeg' } })).toMatchSnapshot()
      })

      test('non-static URL\'s are prefixed with the language', () => {
        expect(getWrapper({ providerProps: { lang, isStatic }, props: { href: '/non-static/page/' } })).toMatchSnapshot()
      })
    })
  })

  describe('when the provider is not used', () => {
    test('it renders correctly', () => {
      expect(mount(
        <Link href='/'>
          {({ href, isStatic, isRelative }) => (
            <a data={{ isStatic, isRelative }} href={href}>
              A link
            </a>
          )}
        </Link>
      )).toMatchSnapshot()
    })
  })
})
