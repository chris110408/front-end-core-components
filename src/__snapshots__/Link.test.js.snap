// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> it renders correctly with minimal props 1`] = `
<Provider>
  <ForwardRef
    href="/"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="/"
      isStatic={[Function]}
      lang=""
    >
      <a
        href="/"
        lang=""
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when a custom component is used it renders correctly 1`] = `
<Provider
  component={[Function]}
>
  <ForwardRef
    href="/page/"
  >
    <LinkInner
      component={[Function]}
      elRef={null}
      href="/page/"
      isStatic={[Function]}
      lang=""
    >
      <CustomLink
        href="/page/"
        isRelative={true}
        isStatic={false}
        lang=""
      >
        <div
          href="/page/"
          isrelative="true"
          isstatic="false"
          lang=""
        >
          A link
        </div>
      </CustomLink>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the context provider has a lang set /en is not prefixed by the language 1`] = `
<Provider
  lang="en"
>
  <ForwardRef
    href="/en"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="/en"
      isStatic={[Function]}
      lang="en"
    >
      <a
        href="/en"
        lang="en"
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the context provider has a lang set /en/ is not prefixed by the language 1`] = `
<Provider
  lang="en"
>
  <ForwardRef
    href="/en/"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="/en/"
      isStatic={[Function]}
      lang="en"
    >
      <a
        href="/en/"
        lang="en"
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the context provider has a lang set /enter/ is still prefixed by the language 1`] = `
<Provider
  lang="en"
>
  <ForwardRef
    href="/enter/"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="/enter/"
      isStatic={[Function]}
      lang="en"
    >
      <a
        href="/en/enter/"
        lang="en"
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the context provider has a lang set absolute URL's are unaffected 1`] = `
<Provider
  lang="en"
>
  <ForwardRef
    href="https://site.com/page/"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="https://site.com/page/"
      isStatic={[Function]}
      lang="en"
    >
      <a
        href="https://site.com/page/"
        lang="en"
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the context provider has a lang set already language prefixed URL's are unaffected 1`] = `
<Provider
  lang="en"
>
  <ForwardRef
    href="/en/page/"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="/en/page/"
      isStatic={[Function]}
      lang="en"
    >
      <a
        href="/en/page/"
        lang="en"
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the context provider has a lang set it prefixes the URL with the lang 1`] = `
<Provider
  lang="en"
>
  <ForwardRef
    href="/page/"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="/page/"
      isStatic={[Function]}
      lang="en"
    >
      <a
        href="/en/page/"
        lang="en"
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the context provider has a lang set when validation for static URL's are in place non-static URL's are prefixed with the language 1`] = `
<Provider
  isStatic={[Function]}
  lang="en"
>
  <ForwardRef
    href="/non-static/page/"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="/non-static/page/"
      isStatic={[Function]}
      lang="en"
    >
      <a
        href="/en/non-static/page/"
        lang="en"
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the context provider has a lang set when validation for static URL's are in place static URL's are unaffected 1`] = `
<Provider
  isStatic={[Function]}
  lang="en"
>
  <ForwardRef
    href="/images/myimage.jpeg"
  >
    <LinkInner
      component="a"
      elRef={null}
      href="/images/myimage.jpeg"
      isStatic={[Function]}
      lang="en"
    >
      <a
        href="/images/myimage.jpeg"
        lang="en"
      >
        A link
      </a>
    </LinkInner>
  </ForwardRef>
</Provider>
`;

exports[`<Link /> when the provider is not used it renders correctly 1`] = `
<ForwardRef
  href="/"
>
  <LinkInner
    component="a"
    elRef={null}
    href="/"
    isStatic={[Function]}
  >
    <a
      href="/"
    >
      A link
    </a>
  </LinkInner>
</ForwardRef>
`;
